### ISSUE TITLE
Feature Request: Before-apply-all hook?

### ISSUE THREAD

**Issue by acutchin-bitpusher** [gruntwork-io/terragrunt#620]
Our consulting firm has several Terraform-ed clients. We commonly keep AWS CLI configurations and credentials for multiple clients and their environments on our workstations, so the risk of accidentally applying one customer's Terraform configuration to another customer's AWS account is frighteningly real. Ditto the risk of applying a "dev" configuration to a "prod" account.

So I implemented a "before_hook" bash script in the "parent"/top-level tfvars file that reads effective AWS CLI credentials from the shell environment and "aws sts get-caller-identity" and compares them to values from the environment/parent tfvars file. This script is interactive and requires the operator to manually enter account and environment names to confirm that they are applying the configuration to the correct AWS account and environment.

We use the typical "live" layout from gruntwork's documentation.

The before-hook credential check script works great when I cd to a module dir and run "terragrunt apply" there. However, as the list of our top-level modules grows, this will become cumbersome and I will want to execute "terragrunt apply-all" from the parent/environment directory.

Currently, when I do so, terragrunt runs the before_hook script in each of the top-level-module directories, in reverse dependency order, frequently in parallel. This requires the operator to enter the confirmations multiple times, and breaks when modules are applied in parallel.

Could we get a "before_all_hook" that runs once, before terragrunt begins iterating through the sub-module dirs and running terraform commands in each?

**Comment by brikis98** [comment]
Seems like a good feature to have. PR to add it welcome!

BTW, this is another layer of sanity checks you can put in to make sure you auth to the right account: https://www.terraform.io/docs/providers/aws/#allowed_account_ids

**Comment by tmeckel** [comment]
@acutchin-bitpusher I've provided a pull request (#680) which supports late interpolation and manipulation of the Terragrunt runtime environment. Could this PR fix any of the issues you described above?

**Comment by acutchin-bitpusher** [comment]
I don't think so. #680 seems to only provide a pre-hook-script re-initialization of the Terragrunt namespace. It doesn't seem to offer any feature that is designed to, or could be exploited to conditionally run the hook script.
